File: statsig-node/src/lib/error_boundary.ts
export class ErrorBoundary {
  static wrap(instance: unknown): void {
    try {
      const obj = instance as Record<string, unknown>;

      _getAllInstanceMethodNames(obj).forEach((name) => {
        const original = obj[name] as (...args: unknown[]) => unknown;
        if ('$EB' in original) {
          return;
        }

        obj[name] = (...args: unknown[]) => {
          return this._capture(name, () => original.apply(instance, args));
        };
        (obj[name] as { $EB: boolean }).$EB = true;
      });
    } catch (err) {
      this._onError('eb:wrap', err);
    }
  }

  private static _capture(tag: string, task: () => unknown): unknown {
    try {
      const res = task();
      if (res && res instanceof Promise) {
        return res.catch((err) => this._onError(tag, err));
      }
      return res;
    } catch (error) {
      this._onError(tag, error);
      return null;
    }
  }

  private static _onError(tag: string, error: unknown): void {
    _tryConvertInvalidArgError(error);

    const printableTag = tag.replace('__INTERNAL_', '');
    console.error('Statsig::' + printableTag, error);
  }
}

function _getAllInstanceMethodNames(
  instance: Record<string, unknown>,
): string[] {
  const names = new Set<string>();

  let proto = Object.getPrototypeOf(instance) as Record<string, unknown>;
  while (proto && proto !== Object.prototype) {
    Object.getOwnPropertyNames(proto)
      .filter((prop) => typeof proto?.[prop] === 'function')
      .forEach((name) => names.add(name));
    proto = Object.getPrototypeOf(proto) as Record<string, unknown>;
  }

  return Array.from(names);
}

function _tryConvertInvalidArgError(error: unknown): void {
  if (typeof error !== 'object' || error === null) {
    return;
  }

  if (!('code' in error) || error.code !== 'InvalidArg') {
    return;
  }

  if (!('message' in error) || typeof error.message !== 'string') {
    return;
  }

  if (
    error.message.startsWith(
      'Failed to recover `StatsigUser` type from napi value',
    )
  ) {
    error.message =
      'Expected StatsigUser instance, plain javascript object is not supported. Please create a StatsigUser instance using `new StatsigUser(...)` instead.';
    return;
  }
}

File: statsig-pyo3/py_src/statsig_python_core/error_boundary.py
class ErrorBoundary:
    @staticmethod
    def wrap(instance):
        try:
            for name in get_all_instance_method_names(instance):
                original = getattr(instance, name)

                if hasattr(original, '_error_boundary_wrapped'):
                    continue

                def create_wrapped(method_name, orig_method):
                    def wrapped(*args, **kwargs):
                        return ErrorBoundary._capture(method_name, lambda: orig_method(*args, **kwargs))

                    wrapped._error_boundary_wrapped = True
                    return wrapped

                setattr(instance, name, create_wrapped(name, original))

        except Exception as err:
            ErrorBoundary._on_error('eb:wrap', err)

    @staticmethod
    def _capture(tag, task):
        try:
            return task()
        except Exception as error:
            ErrorBoundary._on_error(tag, error)
            return None

    @staticmethod
    def _on_error(tag, error):
        print(f"Statsig SDK Error (Python Bindings): {tag}", error)


def get_all_instance_method_names(instance):
    names = set()

    cls = instance.__class__
    classes = [cls]

    parent = cls.__base__
    while parent is not object:
        classes.append(parent)
        parent = parent.__base__

    for cls in classes:
        for name, value in cls.__dict__.items():
            if callable(value) and not name.startswith('__'):
                names.add(name)

    return list(names)
